### PR TITLE
CP-49677 CP-49647 Use URI for improved IPv6 in http.ml, newcli.ml

### PR DESCRIPTION
**Pull request by lindig**
We want to get away from string-based URL handling and delegate this to URI and get correct IPv6 handling along the way.

### CODE REVIEW

File: ocaml/libs/http-lib/http.ml [hunk 1]
```diff
@@ -934,81 +934,67 @@ module Url = struct
...
-          with Not_found -> (
-            try (* ...] *)
-                x |> sub_after ':'
-            with Not_found -> (* ...:port *)
-                              ""
-          )
-        )
+    debug "%s: %s" __FUNCTION__ url ;
+    let fail fmt = Printf.ksprintf failwith fmt in
+    let query = function k, v :: _ -> (k, v) | k, [] -> (k, "") in
```
Comment (psafont) [inline ocaml/libs/http-lib/http.ml:939]: The previous implementation keeps all values for a key, is it significant, though?

File: ocaml/libs/http-lib/http.ml [hunk 2]
```diff
@@ -934,81 +934,67 @@ module Url = struct
...
+                host= Uri.host uri |> Option.get
+              ; auth=
+                  ( match (Uri.user uri, Uri.password uri) with
+                  | Some user, Some pw ->
+                      Some (Basic (user, pw))
+                  | Some user, None ->
+                      Some (Basic (user, ""))
+                  | _ ->
+                      None
+                  )
```
Comment (psafont) [inline ocaml/libs/http-lib/http.ml:963]: This code is duplicated for https as well, can this be moved to a function?

File: ocaml/libs/http-lib/http.ml (hunk 1 above)
Comment (lindig) [reply ocaml/libs/http-lib/http.ml:939]: The URI module can represent a key with multiple values but our representation would have to use duplicate keys - which I think is very questionable and something we should avoid. Currently tests are passing, suggesting we don't use multiple values/keys.

File: ocaml/libs/http-lib/http.ml [hunk 3]
```diff
@@ -934,81 +934,67 @@ module Url = struct
...
+                  | Some user, None ->
+                      debug "%s: user=%s" __FUNCTION__ user ;
+                      Some (Basic (user, ""))
+                  | _ ->
+                      debug "%s: user=none" __FUNCTION__ ;
+                      None
+                  )
+              ; port= Uri.port uri
+              ; ssl= true
+              }
```
Comment (robhoes) [inline ocaml/libs/http-lib/http.ml]: This is almost the same as the `"http"` case above. The only difference is the `ssl` key and additional logging. Can we factor it out?

File: ocaml/libs/http-lib/http.ml [hunk 4]
```diff
@@ -934,81 +934,67 @@ module Url = struct
...
+          (scheme, {data with uri= "/"})
       | _ ->
-          failwith
-            (Printf.sprintf
-               "Failed to parse username password host and port: %s" x
-            )
-    in
-    let reconstruct_uri uri = "/" ^ String.concat "/" uri in
-    let data_of_uri uri =
-      let uri, params = parse_uri (reconstruct_uri uri) in
```
[the remark this answers is not in the document]
Comment (lindig) [reply ocaml/libs/http-lib/http.ml:992]: still used; should be removed

File: ocaml/libs/http-lib/http.ml [hunk 5]
```diff
@@ -1018,27 +979,31 @@ module Url = struct
...
-        let auth =
-          match h.auth with
-          | Some (Basic (username, password)) ->
-              Printf.sprintf "%s:%s" username password |> Option.some
-          | _ ->
-              Option.none
-        in
-        Uri.(
-          make
+  let to_string scheme =
```
Comment (robhoes) [inline ocaml/libs/http-lib/http.ml:982]: `scheme` here isn't of the same type as in `of_string` above, which is a little confusing. The type here is `type t = scheme * data`.

**Review by robhoes — APPROVED** [review]

**Review by psafont — APPROVED** [review]
We should change the base type of Url, in the meantime this adapter is beneficial for parsing urls

**Pull request merged**